core/src/main/java/com/digitalpebble/stormcrawler/protocol/selenium/SeleniumGridProtocol.java

```diff
@@ -0,0 +1,194 @@
+        RemoteWebDriver driver;
+        while ((driver = getDriver()) == null) {}
+        try {
+            // This will block for the page load and any
+            // associated AJAX requests
+            driver.get(url);
+
+            String u = driver.getCurrentUrl();
+
+            // call the filters
+            ProtocolResponse response = filters.filter(driver, metadata);
+            if (response != null) {
+                return response;
+            }
+
+            // if the URL is different then we must have hit a redirection
+            if (!u.equalsIgnoreCase(url)) {
+                byte[] content = new byte[] {};
+                Metadata m = new Metadata();
+                m.addValue(HttpHeaders.LOCATION, u);
+                return new ProtocolResponse(content, 307, m);
+            }
+
+            // if no filters got triggered
+            byte[] content = driver.getPageSource().getBytes();
+            return new ProtocolResponse(content, 200, new Metadata());
+
+        } catch (Exception e) {
+            if (e.getMessage() != null) {
+                if ((e.getMessage().contains("ERR_NAME_NOT_RESOLVED")
+                        || e.getMessage().contains("ERR_CONNECTION_REFUSED")
+                        || e.getMessage().contains("ERR_CONNECTION_CLOSED")
+                        || e.getMessage().contains("ERR_SSL_PROTOCOL_ERROR")
+                        || e.getMessage().contains("ERR_CONNECTION_RESET")
+                        || e.getMessage().contains("ERR_SSL_VERSION_OR_CIPHER_MISMATCH")
+                        || e.getMessage().contains("ERR_ADDRESS_UNREACHABLE"))) {
+                    LOG.info(
+                            "Exception is of webpage related hence continuing with the driver and adding it back to queue");
+                } else {
+                    LOG.error(
+                            "Exception wile doing operation via driver url {} with driver hashcode {}"
+                                    + "with excepiton {}",
+                            url,
+                            driver.hashCode(),
+                            e);
+                    closeConnectionGracefully(driver);
+                    driver = null;
+                }
+            }
+            throw new Exception(e);
+        } finally {
+            // finished with this driver - return it to the queue
+            if (driver != null) driversQueue.put(new Holder(driver, System.currentTimeMillis()));
+        }
+    }
+
+    private final RemoteWebDriver getDriver() {
+        try {
+            return driversQueue.take().getDriver();
+        } catch (Exception e) {
+            return null;
+        }
+    }
+
+    private void closeConnectionGracefully(RemoteWebDriver driver) {
+        try {
+            LOG.info("Before disposing driver : {}", driver.hashCode());
+
+            if (driver.getSessionId() != null) {
```

Comment:
Why are we checking twice? If `driver.getSessionId() != null` the `toString()` should be not null as well?